blog.md

```diff
@@ -1,77 +1,149 @@
+  ```
+  apiVersion: cluster.open-cluster-management.io/v1beta1
+  kind: Placement
+  metadata:
+    name: placement
+    namespace: ns1
+  spec:
+    numberOfClusters: 3
+    prioritizerPolicy:
+      mode: Exact
+      configurations:
+        - scoreCoordinate:
+            type: AddOn
+            addOn:
+              resourceName: default
+              scoreName: cpuAvailable
+          weight: 1
+  ```
+* In placement, if the end user defines the scoreCoordinate type as AddOn, placement will get `AddOnPlacementScore` resource with name "default" in each cluster's namespace, read score "cpuAvailable" in the score list, and use that score to sort clusters.
 
-通过`kubectl describe node` 可以得到node的两个重要属性：Capacity和Allocatable
+You can refer to the [enhancements](https://github.com/open-cluster-management-io/enhancements/blob/main/enhancements/sig-architecture/32-extensiblescheduling/32-extensiblescheduling.md) to learn more details about the design. In the design, how to maintain the lifecycle (create/update/delete) of the `AddOnPlacementScore` CRs is not covered, as we expect the 3rd party controller itself to manage it. And in this blog, we will use an example to show you how to implement a 3rd part controller to update your own scores and extend the multiple clusters scheduling capability with your own scores.
 
 
-![image.png](https://p1-juejin.byteimg.com/tos-cn-i-k3u1fbpfcp/ce5e8c7affe847148af78618664c535e~tplv-k3u1fbpfcp-watermark.image?)
+## How to implement a 3rd part controller
 
-通过对[官网文档](https://kubernetes.io/docs/tasks/administer-cluster/reserve-compute-resources/)的学习，可以知道Allocatable is defined as the amount of compute resources that are available for pods.
+In the OCM environment, you can use the [addon-freamwork](https://github.com/open-cluster-management-io/addon-framework) framework to quickly develop an addon plugin. Here we use it to develop an addon with a hub-agent architecture.
 
-如果cluster只存在一个node，那么使用node.Allocatable 减去 所有当前cluster中所有pod使用的资源，就是可用量。如果cluster中存在多个node，可以得到如下公式，mem同理。
+To develop a controller, you need to consider the following three points.
 
-`cluster_cpu_avaiable = sum(node.status.Allocatable.cpu) - sum(pod_cpu_usage)`
-`cluster_mem_avaiable = sum(node.status.Allocatable.mem) - sum(pod_mem_usage)`
+The following is a specific [example](https://github.com/JiahaoWei-RH/addonplacementscore_collect) to show how to develop it. The main function of this example is which can collect resource usage information on the cluster, and generate an addonplacementscore under the cluster namespace of the hub, and then the placement uses addonplacementscore to select the cluster.
 
-从下面的图片来说，剩余蓝色部分代表的就是计算出来的可用量。
+### 1. Where to run the 3rd party controller
 
+The 3rd part controller could run on either hub or managed cluster. Combined with User Stories, you should be able to distinguish whether the controller should be placed in a hub or a managed cluster.
```

Comment:
For example:

- On the hub, you can get all the clusters metrics (eg. cluster allocatable memory) from Thanos. And you want to generate scores for each cluster based on those metrics, then the 3rd party controller could run on the hub.

- On each managed cluster, you want to monitor the resource usage (eg. CPU ratio) of the cluster and generate a score for each cluster, then the 3rd party controller could run on the managed clusters.